llvm/lib/Passes/PassBuilder.cpp

```diff
@@ -1963,6 +1965,9 @@ Error PassBuilder::parsePassPipeline(ModulePassManager &MPM,
       Pipeline = {{"cgscc", std::move(*Pipeline)}};
     } else if (isFunctionPassName(FirstName,
                                   FunctionPipelineParsingCallbacks)) {
+      if (FirstName == "codegenprepare") {
+        MPM.addPass(RequireAnalysisPass<ProfileSummaryAnalysis, Module>());
+      }
```

Comment:
Should avoid pre-adding dependent passes based on the name, does getResult work?

Reply:
No, I don't think getResult will work. I think PSI needs to be pre-cached using `RequireAnalysisPass<ProfileSummaryAnalysis, Module>()` in the pipeline, because we cannot request a result from an outer-scope analysis. I was having a hard time finding the right place to conditionally require ProfileSummaryAnalysis in the pipeline, so I just used this hack here. Any suggestions would be appreciated.

Reply:
I vaguely recall seeing this hack in another patch by @paperchalice. For now I would drop the FirstName check and just unconditionally add it?

Reply:
Thank you for the advice, the suggestions really helped.